sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -144,16 +144,10 @@ class CacheManager extends Logging {
       } else {
         _.sameResult(plan)
       }
-    val plansToUncache = mutable.Buffer[CachedData]()
-    readLock {
-      val it = cachedData.iterator()
-      while (it.hasNext) {
-        val cd = it.next()
-        if (shouldRemove(cd.plan)) {
-          plansToUncache += cd
-        }
-      }
+    val cachedDataCopy = readLock {
+      cachedData.asScala.clone()
     }
+    val plansToUncache = cachedDataCopy.filter(cd => shouldRemove(cd.plan))
```

Comment:
It seems we always use `whole` loops when performance should be a concern:
https://github.com/databricks/scala-style-guide#traversal-and-zipwithindex
No actual performance impact between the two pattens on your heavy workload?

Reply:
I suspect that it's the logic in `shouldRemove` that takes the time here, and can be done without the lock.

Reply:
Yes, the problem is that the "shouldRemove" function is passed into this call.  If that call is expensive, it causes the lock to be held for arbitrarily long amounts of time.

"shouldRemove" when called from "recacheByPath" causes a full traversal of the entire logical plan tree for every cached plan.  In the process of doing this it will regenerate path strings for every file referenced by every single plan.  In our situations at least this is easily many orders of magnitude more memory overhead than a shallow copy of the list.